src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,186 @@
+        #define FIRMWARE_LOCATIONS "/usr/share/ovmf/ovmf_code_ia32.bin", "/usr/share/edk2/ovmf-ia32/OVMF_CODE.fd"
+#elif __aarch64__
+        #define FIRMWARE_LOCATIONS "/usr/share/AAVMF/AAVMF_CODE.fd"
+#elif __arm__
+        #define FIRMWARE_LOCATIONS "/usr/share/AAVMF/AAVMF32_CODE.fd"
+#else
+        #define FIRMWARE_LOCATIONS
+#endif
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                if (access(firmware, F_OK) == 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.secboot.fd", \
+                "/usr/share/qemu/OVMF_CODE.secboot.fd", \
+                "/usr/share/ovmf/OVMF.secboot.fd", \
+                "/usr/share/OVMF/OVMF_CODE.secboot.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                if (access(firmware, F_OK) == 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 0;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.fd", \
+                "/usr/share/qemu/OVMF_CODE.fd", \
+                "/usr/share/ovmf/OVMF.fd", \
+                "/usr/share/OVMF/OVMF_CODE.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                if (access(firmware, F_OK) == 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        return -ENOENT;
+}
+
+int find_qemu_binary(char **ret_qemu_binary) {
+        int r;
+
+        /*
+         * On success the path to the qemu binary will be stored in `req_qemu_binary`
+         *
+         * If the qemu binary cannot be found -ENOENT will be returned.
+         * If the native architecture is not supported by qemu -ESRCH will be returned;
+         */
+
+        static const char *architecture_to_qemu_table[_ARCHITECTURE_MAX] = {
+                [ARCHITECTURE_ARM64]       = "aarch64",     /* differs from our name */
+                [ARCHITECTURE_ARM]         = "arm",
+                [ARCHITECTURE_ALPHA]       = "alpha",
+                [ARCHITECTURE_X86_64]      = "x86_64",      /* differs from our name */
+                [ARCHITECTURE_X86]         = "i386",        /* differs from our name */
+                [ARCHITECTURE_LOONGARCH64] = "loongarch64",
+                [ARCHITECTURE_MIPS64_LE]   = "mips",        /* differs from our name */
+                [ARCHITECTURE_MIPS_LE]     = "mips",        /* differs from our name */
```

Comment:
are you sure this is right? for us "mips" is mips BE, and "mips-le" is mips LE. looking at this i get the sense that the endianess is swapped here...

Reply:
I've been treating mkosi as the ground truth for logic around this, mkosi refers to mips-le as mips:
https://github.com/systemd/mkosi/blob/814689a613276b74942c8199c7c64b8416118436/mkosi/architecture.py#L103